ludwig/schema/config_object.py

```diff
@@ -0,0 +1,320 @@
+    INPUT_FEATURES,
+    LOSS,
+    MODEL_ECD,
+    MODEL_GBM,
+    MODEL_TYPE,
+    NAME,
+    NUMBER,
+    OPTIMIZER,
+    OUTPUT_FEATURES,
+    PREPROCESSING,
+    PROC_COLUMN,
+    SEQUENCE,
+    SPLIT,
+    TIED,
+    TRAINER,
+    TYPE,
+)
+from ludwig.features.feature_utils import compute_feature_hash
+from ludwig.modules.loss_modules import get_loss_cls
+from ludwig.schema.combiners.base import BaseCombinerConfig
+from ludwig.schema.combiners.concat import ConcatCombinerConfig
+from ludwig.schema.combiners.utils import combiner_registry
+from ludwig.schema.decoders.utils import get_decoder_cls
+from ludwig.schema.defaults.defaults import DefaultsConfig
+from ludwig.schema.encoders.base import PassthroughEncoderConfig
+from ludwig.schema.encoders.binary_encoders import BinaryPassthroughEncoderConfig
+from ludwig.schema.encoders.utils import get_encoder_cls
+from ludwig.schema.features.utils import input_type_registry, output_type_registry
+from ludwig.schema.optimizers import get_optimizer_cls
+from ludwig.schema.preprocessing import PreprocessingConfig
+from ludwig.schema.split import get_split_cls
+from ludwig.schema.trainer import BaseTrainerConfig, ECDTrainerConfig, GBMTrainerConfig
+from ludwig.schema.utils import convert_submodules
+
+DEFAULTS_MODULES = {NAME, COLUMN, PROC_COLUMN, TYPE, TIED, DEFAULT_VALIDATION_METRIC}
+
+
+class BaseFeatureContainer:
+    """Base Feature container for input and output features."""
+
+    def to_dict(self):
+        """Method for getting a dictionary representation of the input features.
+
+        Returns:
+            Dictionary of input features specified.
+        """
+        return convert_submodules(self.__dict__)
+
+    def to_list(self):
+        """Method for getting a list representation of the input features.
+
+        Returns:
+            List of input features specified.
+        """
+        return list(convert_submodules(self.__dict__).values())
+
+
+class InputFeaturesContainer(BaseFeatureContainer):
+    """InputFeatures is a container for all input features."""
+
+    pass
+
+
+class OutputFeaturesContainer(BaseFeatureContainer):
+    """OutputFeatures is a container for all output features."""
+
+    pass
+
+
+class Config:
+    """This class is the implementation of the config object that replaces the need for a config dictionary
+    throughout the project."""
+
+    def __init__(self, config_dict):
+
+        self.model_type = MODEL_ECD
+        self.input_features: InputFeaturesContainer = copy.deepcopy(InputFeaturesContainer())
+        self.output_features: OutputFeaturesContainer = copy.deepcopy(OutputFeaturesContainer())
+        self.combiner: BaseCombinerConfig = copy.deepcopy(ConcatCombinerConfig())
+        self.trainer: BaseTrainerConfig = copy.deepcopy(ECDTrainerConfig())
+        self.preprocessing: PreprocessingConfig = copy.deepcopy(PreprocessingConfig())
+        self.hyperopt = {}
+        self.defaults: DefaultsConfig = copy.deepcopy(DefaultsConfig())
```

Comment:
why deepcopy?

Reply:
Crazy shared state issues between test runs were causing failures when the whole test suite was run but not on individual runs. I believe it has to do with shared pointers for the nested configs throughout the config object.